security/rbac/engine/engine_test.go

```diff
@@ -0,0 +1,264 @@
+// +build go1.10
+
+/*
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package engine
+
+import (
+	"reflect"
+	"sort"
+	"strings"
+	"testing"
+
+	pb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2"
+	cel "github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types/ref"
+	interpreter "github.com/google/cel-go/interpreter"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type programMock struct {
+	out ref.Val
+	err error
+}
+
+func (mock programMock) Eval(vars interface{}) (ref.Val, *cel.EvalDetails, error) {
```

Comment:
IMO, ideally mocking Eval(...) means we will try to do whatever this function will do in actual scenarios. For example, CEL's Eval will only accept map type or Activation type, so we can implement the similar logic here: only when the vars is of the two types will we return a successful evaluation.
Of course, since this is a mock, we can assume Eval will return  nil val if vars is a string of "runtime error", and non-nil if vars is a string of "unknown error". The point is, setting field directly inside the mock struct doesn't quite fit the purpose of "mocking" objects.

Reply:
Follow up from offline conversation: the current testing structure functionally serves its purpose, so this is just a preference of coding style. If other reviewers have strong opinions on this, I can redesign it.